clang/include/clang/AST/Type.h

```diff
@@ -4429,6 +4433,210 @@ class FunctionNoProtoType : public FunctionType, public llvm::FoldingSetNode {
+             FE_ExcludeThreadLocalVars;
+    case Kind::NonAllocating:
+      // Same as NonBlocking, except without FE_ExcludeStaticLocalVars
+      return FE_InferrableOnCallees | FE_ExcludeThrow | FE_ExcludeCatch |
+             FE_ExcludeObjCMessageSend | FE_ExcludeThreadLocalVars;
+    case Kind::None:
+      break;
+    }
+    llvm_unreachable("unknown effect kind");
+  }
+
+  /// The description printed in diagnostics, e.g. 'nonblocking'.
+  StringRef name() const;
+
+  /// Return true if adding or removing the effect as part of a type conversion
+  /// should generate a diagnostic.
+  bool shouldDiagnoseConversion(bool Adding, QualType OldType,
+                                const FunctionEffectSet &OldFX,
+                                QualType NewType,
+                                const FunctionEffectSet &NewFX) const;
+
+  /// Return true if adding or removing the effect in a redeclaration should
+  /// generate a diagnostic.
+  bool shouldDiagnoseRedeclaration(bool Adding, const FunctionDecl &OldFunction,
+                                   const FunctionEffectSet &OldFX,
+                                   const FunctionDecl &NewFunction,
+                                   const FunctionEffectSet &NewFX) const;
+
+  /// Return true if adding or removing the effect in a C++ virtual method
+  /// override should generate a diagnostic.
+  OverrideResult
+  shouldDiagnoseMethodOverride(bool Adding, const CXXMethodDecl &OldMethod,
+                               const FunctionEffectSet &OldFX,
+                               const CXXMethodDecl &NewMethod,
+                               const FunctionEffectSet &NewFX) const;
+
+  /// Return true if the effect is allowed to be inferred on the callee,
+  /// which is either a FunctionDecl or BlockDecl.
+  /// This is only used if the effect has FE_InferrableOnCallees flag set.
+  /// Example: This allows nonblocking(false) to prevent inference for the
+  /// function.
+  bool canInferOnFunction(const Decl &Callee) const;
+
+  // Return false for success. When true is returned for a direct call, then the
+  // FE_InferrableOnCallees flag may trigger inference rather than an immediate
+  // diagnostic. Caller should be assumed to have the effect (it may not have it
+  // explicitly when inferring).
+  bool shouldDiagnoseFunctionCall(bool Direct,
+                                  const FunctionEffectSet &CalleeFX) const;
+
+  friend bool operator==(const FunctionEffect &LHS, const FunctionEffect &RHS) {
+    return LHS.FKind == RHS.FKind;
+  }
+  friend bool operator!=(const FunctionEffect &LHS, const FunctionEffect &RHS) {
+    return LHS.FKind != RHS.FKind;
+  }
+  friend bool operator<(const FunctionEffect &LHS, const FunctionEffect &RHS) {
+    return LHS.FKind < RHS.FKind;
+  }
+};
+
+/// A value type, representing a set of FunctionEffects. To reduce superfluous
+/// retain/release, however, prefer to pass by reference.
+class FunctionEffectSet {
```

Comment:
This seems less overly elaborate now, without the uniqued sets and without a separate `MutableFunctionEffectSet`.

I'm trying to decide whether `FunctionEffectSet` should hold the "computed effects" too, or whether `FunctionProtoType` should hold that separately.

In any case, something other than `FunctionProtoType` needs to handle merging the computed effects (once their expressions are non-dependent), because that operation would use `Sema::PerformContextuallyConvertToBool()`, and `FunctionProtoType` doesn't have a `Sema&` or precedent for using one.

Reply:
> I'm trying to decide whether `FunctionEffectSet` should hold the "computed effects" too, or whether `FunctionProtoType` should hold that separately.

If it keeps the `FunctionProtoType` smaller/simpler then I’d put it in the `FunctionEffectSet`, otherwise I don’t think it matters too much.

> In any case, something other than `FunctionProtoType` needs to handle merging the computed effects (once their expressions are non-dependent), because that operation would use `Sema::PerformContextuallyConvertToBool()`, 

Yeah, typically Sema creates the AST nodes, so it makes sense to put e.g. `MergeFunctionEffects` or whatever in Sema.

> and `FunctionProtoType` doesn't have a `Sema&` or precedent for using one.

Yeah, typically we put functions like that in Sema (or in the ASTContext in some cases) and have Sema operate on the types rather than the other way around. AST nodes are mostly ‘data’; the actual ‘logic’ is typically somewhere else, if you will (most of the member functions of AST nodes just deal w/ accessing data/information that’s part of the node).

Reply:
Thanks for all this. Now I think that `FunctionProtoType` can just hold 1+ `FunctionEffects` and 1+ `ComputedFunctionEffects` and return those in something like `FunctionEffectSetRef`, holding `ArrayRef`s. That can be responsible for the set logic. `FunctionEffectSet` could either duplicate that or derive from it while holding mutable vectors...

Reply:
> Now I think that `FunctionProtoType` can just hold trailing `FunctionEffect`s and `ComputedFunctionEffect`s

Yeah, to me, that also looks like it’d be the right approach here.

Reply:
Done